diff --git a/lib/dotgen/mincross.c b/lib/dotgen/mincross.c
index bdd32ea5f..c9a1d022a 100644
--- a/lib/dotgen/mincross.c
+++ b/lib/dotgen/mincross.c
@@ -1513,3 +1513,3 @@ static int local_cross(elist l, int dir) {
 static int64_t rcross(graph_t *g, int r) {
-  int top, bot, max, i, k;
+  int top, bot, max, i;
   node_t **rtop, *v;
@@ -1520,3 +1520,4 @@ static int64_t rcross(graph_t *g, int r) {
 
-  int *Count = gv_calloc(GD_rank(Root)[r + 1].n + 1, sizeof(int));
+  const int count_size = GD_rank(Root)[r + 1].n + 1;
+  int *Count = gv_calloc(count_size, sizeof(int));
 
@@ -1526,4 +1527,8 @@ static int64_t rcross(graph_t *g, int r) {
       for (i = 0; (e = ND_out(rtop[top]).list[i]); i++) {
-        for (k = ND_order(aghead(e)) + 1; k <= max; k++)
-          cross += Count[k] * ED_xpenalty(e);
+        int after = 0;
+        for (int k = max + 1; k > 0; k -= k & -k)
+          after += Count[k];
+        for (int k = ND_order(aghead(e)) + 1; k > 0; k -= k & -k)
+          after -= Count[k];
+        cross += after * ED_xpenalty(e);
       }
@@ -1534,3 +1539,4 @@ static int64_t rcross(graph_t *g, int r) {
         max = inv;
-      Count[inv] += ED_xpenalty(e);
+      for (int k = inv + 1; k < count_size; k += k & -k)
+        Count[k] += ED_xpenalty(e);
     }
